pkg/controllers/cloud/cloudprovideraccount_controller.go

```diff
@@ -215,25 +240,107 @@ func (r *CloudProviderAccountReconciler) startPollingThread(namespacedName *type
 
 func (r *CloudProviderAccountReconciler) addAccountProviderType(namespacedName *types.NamespacedName,
 	provider runtimev1alpha1.CloudProvider) common.ProviderType {
-	r.mutex.Lock()
-	defer r.mutex.Unlock()
-
 	providerType := common.ProviderType(provider)
 	r.accountProviderType[*namespacedName] = providerType
-
 	return providerType
 }
 
 func (r *CloudProviderAccountReconciler) removeAccountProviderType(namespacedName *types.NamespacedName) {
-	r.mutex.Lock()
-	defer r.mutex.Unlock()
-
 	delete(r.accountProviderType, *namespacedName)
 }
 
 func (r *CloudProviderAccountReconciler) getAccountProviderType(namespacedName *types.NamespacedName) common.ProviderType {
-	r.mutex.Lock()
-	defer r.mutex.Unlock()
-
 	return r.accountProviderType[*namespacedName]
 }
+
+// getCpaBySecret returns nil only when the Secret is not used by any CloudProvideAccount CR,
+// otherwise the dependent CloudProvideAccount CRs will be returned.
+func (r *CloudProviderAccountReconciler) getCpaBySecret(s types.NamespacedName) (error, []crdv1alpha1.CloudProviderAccount) {
+	cpaList := &crdv1alpha1.CloudProviderAccountList{}
+	if err := r.Client.List(context.TODO(), cpaList, &client.ListOptions{}); err != nil {
+		return fmt.Errorf("failed to get CloudProviderAccount list, err %v", err), nil
+	}
+	var cpaItems []crdv1alpha1.CloudProviderAccount
+	for _, cpa := range cpaList.Items {
+		if cpa.Spec.AWSConfig != nil {
+			if cpa.Spec.AWSConfig.SecretRef.Name == s.Name &&
+				cpa.Spec.AWSConfig.SecretRef.Namespace == s.Namespace {
+				cpaItems = append(cpaItems, cpa)
+			}
+		}
+
+		if cpa.Spec.AzureConfig != nil {
+			if cpa.Spec.AzureConfig.SecretRef.Name == s.Name &&
+				cpa.Spec.AzureConfig.SecretRef.Namespace == s.Namespace {
+				cpaItems = append(cpaItems, cpa)
+			}
+		}
+	}
+	return nil, cpaItems
+}
+
+// watchSecret watch the Secret objects.
+func (r *CloudProviderAccountReconciler) watchSecret() {
+	for {
+		event, ok := <-r.watcher.ResultChan()
+		if !ok {
+			r.resetSecretWatcher()
+		} else {
+			switch event.Type {
+			case watch.Modified:
+				r.mutex.Lock()
+				secret := event.Object.(*v1.Secret)
+				namespacedName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
+				r.Log.WithName("Secret").Info("Received request", "Secret", namespacedName, "operation", "update")
+				err, cpaItems := r.getCpaBySecret(namespacedName)
+				if err != nil {
+					r.Log.WithName("Secret").Error(err, "error getting CPA by Secret", "account", namespacedName)
+				}
+				for _, cpa := range cpaItems {
+					// If the credentials are updated to new cloud account then vpc cache is not updated. As a
+					// result vpc cache will contain vpcs from old and new accounts.
+					accountNamespacedName := types.NamespacedName{Namespace: cpa.Namespace, Name: cpa.Name}
```

Comment:
is it a bug?

Reply:
@Anandkumar26 Please comment on this.

Reply:
Consider we have account 1, with cred 1.  We get 10 VPCs from account.
Now user updates the cred 1 to cred 2. (Cred 2 can potential point to different account, same region). 

In this particular scenario,  VPCs cache will have:
- VPCs from account 1 (Since they never got deleted upon secret modification)
- VPCs from new account (updated credentials)

Reply:
@Nithish555 please remove the comment..

Reply:
Please remove the entire comment